### PR TITLE
storage.onChanged clarrification

### PR DESCRIPTION
**Pull request by rebloor**
#### Summary
Clarifies the behavior of and data returned by storage.onChanged. Provides the content requested in [Bug 1621162](https://bugzilla.mozilla.org/show_bug.cgi?id=1621162).

#### Metadata

This PR…
- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 1]
```diff
@@ -15,7 +15,9 @@ browser-compat: webextensions.api.storage.onChanged
 ---
 {{AddonSidebar()}}
 
-Fired when one or more items change.
+Fired when {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} executes against a storage area.
```
Comment (rpl) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:18]: @rebloor technically the onChanged event would also be fired in response to calls to the `remove` and `clear` methods, and so we should mention them here along with `set`.

I also confirmed that the clarification paragraph that follows (related to the behavior to expect when `set` is called) should be ok as is:

- storage.local: [looking to the implementation for storage.local.set/remove/clear from ExtensionStorageIDB.jsm](https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ExtensionStorageIDB.jsm) seems to confirm (as I recalled) that `remove` and `clear` would only fire the `onChanged` event if a key was actually removed.
- storage.sync:  [storage.sync.set](https://github.com/mozilla/application-services/blob/a074bf651bf3bec56bc0b78ac876888904be1392/components/webext-storage/src/api.rs#L202-L221) should be presenting the same behavior that storage.local.set presents and as already described in the clarification paragraph as is (while `remove` and `clear` are only fired when there were keys removed, as described above for the storage.local implementation).

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 2]
```diff
@@ -15,7 +15,9 @@ browser-compat: webextensions.api.storage.onChanged
...
+
+As this event is triggered by {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}}, it's possible to receive a callback when there is no change to the underlying data. Also, the information returned includes all keys within the storage area {{WebExtAPIRef('storage.StorageArea.set','storageArea.set')}} ran against. Your code needs to determine what changes have occurred by examining the content of the `changes` argument.
```
Comment (rpl) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:20]: Nit, I'm wondering if we could reword a bit "Your code needs to determine ..." to say something like `The extension may determine what changes have occurred by examining the content of the `changes` argument received by the `onChanged` listeners.`. wdyt?

**Review by rpl — CHANGES_REQUESTED** [review]
@rebloor looks good but there is a small fix to be applied (mentioning also `remove` and `clear` as methods that triggers an onChanged event to be fire) and a small proposed tweak to the clarification paragraph that follows.

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md (hunk 1 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:18]: @rpl can I clarify, do `remove` and `clear` triggers also result in all the keys being returned or just the keys that were removed or cleared?

Comment (rpl) [reply files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md]: > @rpl can I clarify, do remove and clear triggers also result in all the keys being returned or just the keys that were removed or cleared?

They both fire an event only if something has been removed (or cleared), from the perspective of what is included in the changes emitted in an onChanged event:
- on `remove` calls: only the keys removed are expected to be included 
- on `clear` calls: we will remove everything and so the changes emitted will technically include all stored keys (if there wasn't any then onChanged wouldn't be fired).

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md [hunk 3]
```diff
@@ -40,10 +42,10 @@ Events have three functions:
 
 - `callback`
 
-  - : Function that will be called when this event occurs. The function will be passed the following arguments:
+  - : The function called when this event occurs. The function is passed these arguments:
 
     - `changes`
-      - : `object`. Object describing the change. This contains one property for each key that changed. The name of the property is the name of the key that changed, and its value is a {{WebExtAPIRef('storage.StorageChange')}} object describing the change to that item.
+      - : `object`. Object describing the change. This object contains properties for all the keys in the storage area where {{WebExtAPIRef('storage.StorageArea.set','storageAred.set')}} occurred. The name of each property is the name of each key. The value of each key is a {{WebExtAPIRef('storage.StorageChange')}} object describing the change to that item.
```
Comment (rpl) [inline files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md:48]: I just noticed that this is currently saying "all the keys in the storage area where set occurred", but the actual behavior is that the changes object will contains all keys that where included in the `set` call, even if the old and new value may still be the same.

Any other key stored in the same area that was not part of what was passed to the `set` call will still not be part of the `changes` object.

would you mind to tweak that part?

**Review by rpl — CHANGES_REQUESTED** [review]

File: files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md (hunk 3 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/api/storage/onchanged/index.md]: Done

**Review by rpl — APPROVED** [review]
thanks!

**Pull request merged**